File: Ruby/cache_utils/cache_utils.rb
# frozen_string_literal: true

# cache_utils.rb - Memory Cache with LRU Eviction Policy
# 
# A lightweight, zero-dependency memory cache implementation for Ruby
# with TTL support, LRU eviction, and thread safety.
#
# Features:
# - LRU (Least Recently Used) eviction policy
# - TTL (Time To Live) support for cache entries
# - Maximum capacity limits
# - Thread-safe operations
# - Statistics tracking (hits, misses, evictions)
# - Zero external dependencies
#
# Author: AllToolkit
# Date: 2026-04-20

require 'thread'

module CacheUtils
  # Cache entry representing a single cached item
  class Entry
    attr_accessor :key, :value, :created_at, :expires_at, :prev, :next

    def initialize(key:, value:, ttl: nil)
      @key = key
      @value = value
      @created_at = Time.now
      @expires_at = ttl ? @created_at + ttl : nil
      @prev = nil
      @next = nil
    end

    def expired?
      @expires_at && Time.now > @expires_at
    end

    def remaining_ttl
      return nil unless @expires_at
      [@expires_at - Time.now, 0].max
    end

    def age
      Time.now - @created_at
    end
  end

  # LRU Cache implementation with TTL support
  class LRUCache
    attr_reader :max_size, :stats

    def initialize(max_size: 1000, default_ttl: nil)
      @max_size = max_size
      @default_ttl = default_ttl
      @cache = {} # key => Entry
      @head = nil # Most recently used
      @tail = nil # Least recently used
      @mutex = Mutex.new
      reset_stats
    end

    # Get a value from the cache
    def get(key)
      @mutex.synchronize do
        entry = @cache[key]
        
        if entry.nil?
          increment_misses
          return nil
        end

        if entry.expired?
          remove_entry(entry)
          @cache.delete(key)
          increment_misses
          increment_evictions
          return nil
        end

        # Move to front (most recently used)
        move_to_front(entry)
        increment_hits
        entry.value
      end
    end

    # Alias for get
    def [](key)
      get(key)
    end

    # Set a value in the cache
    def set(key, value, ttl: nil)
      @mutex.synchronize do
        ttl ||= @default_ttl

        # If key exists, update and move to front
        if @cache.key?(key)
          entry = @cache[key]
          entry.value = value
          entry.created_at = Time.now
          entry.expires_at = ttl ? entry.created_at + ttl : nil
          move_to_front(entry)
          return value
        end

        # Evict if at capacity
        evict_lru while @cache.size >= @max_size

        # Create new entry
        entry = Entry.new(key: key, value: value, ttl: ttl)
        @cache[key] = entry
        add_to_front(entry)
        value
      end
    end

    # Alias for set
    def []=(key, value)
      set(key, value)
    end

    # Check if key exists and is not expired
    def has_key?(key)
      @mutex.synchronize do
        entry = @cache[key]
        return false if entry.nil?
        
        if entry.expired?
          remove_entry(entry)
          @cache.delete(key)
          return false
        end
        
        true
      end
    end

    # Alias for has_key?
    def key?(key)
      has_key?(key)
    end

    # Delete a key from the cache
    def delete(key)
      @mutex.synchronize do
        entry = @cache[key]
        return nil if entry.nil?
        
        remove_entry(entry)
        @cache.delete(key)
        entry.value
      end
    end

    # Get or compute - fetch from cache or execute block
    def fetch(key, ttl: nil)
      value = get(key)
      return value unless value.nil?

      if block_given?
        value = yield
        set(key, value, ttl: ttl)
        value
      else
        nil
      end
    end

    # Clear all cache entries
    def clear
      @mutex.synchronize do
        @cache.clear
        @head = nil
        @tail = nil
        reset_stats
        true
      end
    end

    # Get current cache size
    def size
      @mutex.synchronize do
        cleanup_expired
        @cache.size
      end
    end

    # Get all keys
    def keys
      @mutex.synchronize do
        cleanup_expired
        @cache.keys
      end
    end

    # Get all values
    def values
      @mutex.synchronize do
        cleanup_expired
        @cache.values.map(&:value)
      end
    end

    # Get cache statistics
    def stats
      @mutex.synchronize do
        @stats.dup
      end
    end

    # Get hit rate (0.0 to 1.0)
    def hit_rate
      @mutex.synchronize do
        total = @stats[:hits] + @stats[:misses]
        return 0.0 if total.zero?
        @stats[:hits].to_f / total
      end
    end

    # Cleanup expired entries
    def cleanup
      @mutex.synchronize do
        cleanup_expired
      end
    end

    # Get entry info (without affecting LRU order)
    def entry_info(key)
      @mutex.synchronize do
        entry = @cache[key]
        return nil if entry.nil?

        {
          key: key,
          value: entry.value,
          age: entry.age.round(2),
          remaining_ttl: entry.remaining_ttl&.round(2),
          expired: entry.expired?
        }
      end
    end

    # Update TTL for an existing key
    def touch(key, ttl: nil)
      @mutex.synchronize do
        entry = @cache[key]
        return false if entry.nil?

        entry.expires_at = ttl ? Time.now + ttl : nil
        move_to_front(entry)
        true
      end
    end

    # Increment a numeric value (atomic)
    def increment(key, delta = 1, ttl: nil)
      @mutex.synchronize do
        entry = @cache[key]
        
        if entry.nil?
          set_internal(key, delta, ttl: ttl)
          return delta
        end

        entry.value = entry.value.to_i + delta
        move_to_front(entry)
        entry.value
      end
    end

    # Decrement a numeric value (atomic)
    def decrement(key, delta = 1, ttl: nil)
      increment(key, -delta, ttl: ttl)
    end

    # Get multiple values at once
    def get_multi(*keys)
      @mutex.synchronize do
        result = {}
        keys.each do |key|
          entry = @cache[key]
          if entry && !entry.expired?
            move_to_front(entry)
            result[key] = entry.value
            increment_hits
          else
            increment_misses
          end
        end
        result
      end
    end

    # Set multiple values at once
    def set_multi(hash, ttl: nil)
      @mutex.synchronize do
        hash.each do |key, value|
          set_internal(key, value, ttl: ttl)
        end
        true
      end
    end

    private

    def set_internal(key, value, ttl: nil)
      ttl ||= @default_ttl

      if @cache.key?(key)
        entry = @cache[key]
        entry.value = value
        entry.created_at = Time.now
        entry.expires_at = ttl ? entry.created_at + ttl : nil
        move_to_front(entry)
        return value
      end

      evict_lru while @cache.size >= @max_size

      entry = Entry.new(key: key, value: value, ttl: ttl)
      @cache[key] = entry
      add_to_front(entry)
      value
    end

    def move_to_front(entry)
      return if entry == @head
      
      remove_entry(entry)
      add_to_front(entry)
    end

    def add_to_front(entry)
      entry.prev = nil
      entry.next = @head

      @head.prev = entry if @head
      @head = entry
      @tail ||= entry
    end

    def remove_entry(entry)
      if entry.prev
        entry.prev.next = entry.next
      else
        @head = entry.next
      end

      if entry.next
        entry.next.prev = entry.prev
      else
        @tail = entry.prev
      end
    end

    def evict_lru
      return unless @tail

      entry = @tail
      remove_entry(entry)
      @cache.delete(entry.key)
      increment_evictions
    end

    def cleanup_expired
      keys_to_delete = []
      @cache.each do |key, entry|
        keys_to_delete << key if entry.expired?
      end

      keys_to_delete.each do |key|
        entry = @cache.delete(key)
        remove_entry(entry) if entry
      end

      keys_to_delete.size
    end

    def reset_stats
      @stats = {
        hits: 0,
        misses: 0,
        evictions: 0
      }
    end

    def increment_hits
      @stats[:hits] += 1
    end

    def increment_misses
      @stats[:misses] += 1
    end

    def increment_evictions
      @stats[:evictions] += 1
    end
  end

  # Simple cache without LRU (faster for small caches)
  class SimpleCache
    attr_reader :max_size, :stats

    def initialize(max_size: 1000, default_ttl: nil)
      @max_size = max_size
      @default_ttl = default_ttl
      @cache = {}
      @mutex = Mutex.new
      reset_stats
    end

    def get(key)
      @mutex.synchronize do
        entry = @cache[key]
        
        if entry.nil?
          increment_misses
          return nil
        end

        if entry[:expires_at] && Time.now > entry[:expires_at]
          @cache.delete(key)
          increment_misses
          increment_evictions
          return nil
        end

        increment_hits
        entry[:value]
      end
    end

    def [](key)
      get(key)
    end

    def set(key, value, ttl: nil)
      @mutex.synchronize do
        ttl ||= @default_ttl

        # Simple FIFO eviction when at capacity
        if @cache.size >= @max_size && !@cache.key?(key)
          @cache.shift # Remove first (oldest) entry
          increment_evictions
        end

        @cache[key] = {
          value: value,
          created_at: Time.now,
          expires_at: ttl ? Time.now + ttl : nil
        }
        value
      end
    end

    def []=(key, value)
      set(key, value)
    end

    def has_key?(key)
      @mutex.synchronize do
        entry = @cache[key]
        return false if entry.nil?
        
        if entry[:expires_at] && Time.now > entry[:expires_at]
          @cache.delete(key)
          return false
        end
        
        true
      end
    end

    def delete(key)
      @mutex.synchronize do
        entry = @cache.delete(key)
        entry&.dig(:value)
      end
    end

    def fetch(key, ttl: nil)
      value = get(key)
      return value unless value.nil?

      if block_given?
        value = yield
        set(key, value, ttl: ttl)
        value
      else
        nil
      end
    end

    def clear
      @mutex.synchronize do
        @cache.clear
        reset_stats
        true
      end
    end

    def size
      @mutex.synchronize { @cache.size }
    end

    def keys
      @mutex.synchronize { @cache.keys }
    end

    def values
      @mutex.synchronize { @cache.values.map { |e| e[:value] } }
    end

    def stats
      @mutex.synchronize { @stats.dup }
    end

    def hit_rate
      total = @stats[:hits] + @stats[:misses]
      return 0.0 if total.zero?
      @stats[:hits].to_f / total
    end

    private

    def reset_stats
      @stats = { hits: 0, misses: 0, evictions: 0 }
    end

    def increment_hits
      @stats[:hits] += 1
    end

    def increment_misses
      @stats[:misses] += 1
    end

    def increment_evictions
      @stats[:evictions] += 1
    end
  end

  # Factory method for creating caches
  def self.create_lru_cache(max_size: 1000, default_ttl: nil)
    LRUCache.new(max_size: max_size, default_ttl: default_ttl)
  end

  def self.create_simple_cache(max_size: 1000, default_ttl: nil)
    SimpleCache.new(max_size: max_size, default_ttl: default_ttl)
  end

  # Global default cache instance
  @default_cache = nil

  def self.default_cache
    @default_cache ||= LRUCache.new
  end

  def self.reset_default_cache!
    @default_cache = nil
  end

  # Convenience methods using default cache
  def self.get(key)
    default_cache.get(key)
  end

  def self.set(key, value, ttl: nil)
    default_cache.set(key, value, ttl: ttl)
  end

  def self.fetch(key, ttl: nil)
    default_cache.fetch(key, ttl: ttl) { yield if block_given? }
  end

  def self.delete(key)
    default_cache.delete(key)
  end

  def self.clear
    default_cache.clear
  end
end

File: Kotlin/cache_utils/CacheUtils.kt
/**
 * Cache Utilities for Kotlin
 * 
 * Provides efficient caching implementations with zero external dependencies.
 * Includes LRU cache, TTL cache, and composite cache with memory management.
 * 
 * @author AllToolkit
 * @date 2026-05-03
 */

package cache_utils

import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentLinkedDeque
import java.util.concurrent.atomic.AtomicLong
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/**
 * LRU (Least Recently Used) Cache implementation
 * Automatically evicts least recently used entries when capacity is reached
 */
class LRUCache<K, V>(
    private val capacity: Int,
    private val onEvict: ((K, V) -> Unit)? = null
) {
    init {
        require(capacity > 0) { "Capacity must be positive" }
    }
    
    private val cache = ConcurrentHashMap<K, V>()
    private val accessOrder = ConcurrentLinkedDeque<K>()
    private val lock = ReentrantLock()
    
    @Volatile
    private var hitCount = AtomicLong(0)
    
    @Volatile
    private var missCount = AtomicLong(0)
    
    /**
     * Get a value from the cache
     * @param key The key to look up
     * @return The value if found, null otherwise
     */
    fun get(key: K): V? {
        return lock.withLock {
            val value = cache[key]
            if (value != null) {
                hitCount.incrementAndGet()
                // Move to front (most recently used)
                accessOrder.remove(key)
                accessOrder.addFirst(key)
                value
            } else {
                missCount.incrementAndGet()
                null
            }
        }
    }
    
    /**
     * Put a value into the cache
     * @param key The key to store
     * @param value The value to store
     * @return The evicted value if any, null otherwise
     */
    fun put(key: K, value: V): V? {
        return lock.withLock {
            // If key exists, remove it from access order first
            if (cache.containsKey(key)) {
                accessOrder.remove(key)
            }
            
            // If at capacity, evict LRU entry
            var evictedValue: V? = null
            while (cache.size >= capacity && accessOrder.isNotEmpty()) {
                val lruKey = accessOrder.removeLast()
                evictedValue = cache.remove(lruKey)
                evictedValue?.let { onEvict?.invoke(lruKey, it) }
            }
            
            cache[key] = value
            accessOrder.addFirst(key)
            evictedValue
        }
    }
    
    /**
     * Remove a key from the cache
     * @param key The key to remove
     * @return The removed value if any, null otherwise
     */
    fun remove(key: K): V? {
        return lock.withLock {
            val value = cache.remove(key)
            if (value != null) {
                accessOrder.remove(key)
            }
            value
        }
    }
    
    /**
     * Check if a key exists in the cache
     */
    fun contains(key: K): Boolean = cache.containsKey(key)
    
    /**
     * Clear all entries from the cache
     */
    fun clear() {
        lock.withLock {
            cache.clear()
            accessOrder.clear()
        }
    }
    
    /**
     * Get the current size of the cache
     */
    val size: Int get() = cache.size
    
    /**
     * Get cache statistics
     */
    fun getStats(): CacheStats {
        val hits = hitCount.get()
        val misses = missCount.get()
        return CacheStats(
            size = size,
            capacity = capacity,
            hits = hits,
            misses = misses,
            hitRate = if (hits + misses > 0) hits.toDouble() / (hits + misses) else 0.0
        )
    }
}

/**
 * TTL (Time-To-Live) Cache implementation
 * Entries automatically expire after a specified duration
 */
class TTLCache<K, V>(
    private val defaultTtlMs: Long = 60_000L, // 1 minute default
    private val maxSize: Int = 1000,
    private val onExpire: ((K, V) -> Unit)? = null
) {
    private data class Entry<V>(
        val value: V,
        val expiresAt: Long
    ) {
        fun isExpired(): Boolean = System.currentTimeMillis() > expiresAt
    }
    
    private val cache = ConcurrentHashMap<K, Entry<V>>()
    private val lock = ReentrantLock()
    
    @Volatile
    private var hitCount = AtomicLong(0)
    
    @Volatile
    private var missCount = AtomicLong(0)
    
    @Volatile
    private var evictionCount = AtomicLong(0)
    
    /**
     * Get a value from the cache
     * @param key The key to look up
     * @return The value if found and not expired, null otherwise
     */
    fun get(key: K): V? {
        return lock.withLock {
            val entry = cache[key]
            if (entry != null) {
                if (entry.isExpired()) {
                    cache.remove(key)
                    evictionCount.incrementAndGet()
                    onExpire?.invoke(key, entry.value)
                    missCount.incrementAndGet()
                    null
                } else {
                    hitCount.incrementAndGet()
                    entry.value
                }
            } else {
                missCount.incrementAndGet()
                null
            }
        }
    }
    
    /**
     * Put a value into the cache with default TTL
     * @param key The key to store
     * @param value The value to store
     */
    fun put(key: K, value: V) {
        put(key, value, defaultTtlMs)
    }
    
    /**
     * Put a value into the cache with custom TTL
     * @param key The key to store
     * @param value The value to store
     * @param ttlMs Time to live in milliseconds
     */
    fun put(key: K, value: V, ttlMs: Long) {
        lock.withLock {
            // Cleanup if approaching max size
            if (cache.size >= maxSize) {
                cleanupExpired()
                if (cache.size >= maxSize) {
                    // Force eviction of oldest entries
                    val toRemove = cache.size - maxSize + 1
                    cache.entries
                        .sortedBy { it.value.expiresAt }
                        .take(toRemove)
                        .forEach { (k, v) ->
                            cache.remove(k)
                            evictionCount.incrementAndGet()
                            onExpire?.invoke(k, v.value)
                        }
                }
            }
            
            cache[key] = Entry(value, System.currentTimeMillis() + ttlMs)
        }
    }
    
    /**
     * Remove a key from the cache
     */
    fun remove(key: K): V? {
        return lock.withLock {
            cache.remove(key)?.value
        }
    }
    
    /**
     * Check if a key exists and is not expired
     */
    fun contains(key: K): Boolean {
        return get(key) != null
    }
    
    /**
     * Clean up all expired entries
     * @return Number of entries removed
     */
    fun cleanupExpired(): Int {
        return lock.withLock {
            var removed = 0
            val iterator = cache.entries.iterator()
            while (iterator.hasNext()) {
                val entry = iterator.next()
                if (entry.value.isExpired()) {
                    iterator.remove()
                    removed++
                    onExpire?.invoke(entry.key, entry.value.value)
                }
            }
            evictionCount.addAndGet(removed.toLong())
            removed
        }
    }
    
    /**
     * Clear all entries from the cache
     */
    fun clear() {
        lock.withLock {
            cache.clear()
        }
    }
    
    /**
     * Get the current size of the cache (including potentially expired entries)
     */
    val size: Int get() = cache.size
    
    /**
     * Get cache statistics
     */
    fun getStats(): TTLCacheStats {
        val hits = hitCount.get()
        val misses = missCount.get()
        val evictions = evictionCount.get()
        return TTLCacheStats(
            size = size,
            maxSize = maxSize,
            hits = hits,
            misses = misses,
            hitRate = if (hits + misses > 0) hits.toDouble() / (hits + misses) else 0.0,
            evictions = evictions,
            defaultTtlMs = defaultTtlMs
        )
    }
}

/**
 * LFU (Least Frequently Used) Cache implementation
 * Evicts entries that are accessed least frequently
 */
class LFUCache<K, V>(
    private val capacity: Int,
    private val onEvict: ((K, V) -> Unit)? = null
) {
    init {
        require(capacity > 0) { "Capacity must be positive" }
    }
    
    private data class Entry<V>(
        val value: V,
        var frequency: Long = 1
    )
    
    private val cache = ConcurrentHashMap<K, Entry<V>>()
    private val lock = ReentrantLock()
    
    @Volatile
    private var hitCount = AtomicLong(0)
    
    @Volatile
    private var missCount = AtomicLong(0)
    
    /**
     * Get a value from the cache
     * Increments access frequency on hit
     */
    fun get(key: K): V? {
        return lock.withLock {
            val entry = cache[key]
            if (entry != null) {
                hitCount.incrementAndGet()
                entry.frequency++
                entry.value
            } else {
                missCount.incrementAndGet()
                null
            }
        }
    }
    
    /**
     * Put a value into the cache
     * Evicts least frequently used entry if at capacity
     */
    fun put(key: K, value: V): V? {
        return lock.withLock {
            // If key exists, update value and reset frequency
            if (cache.containsKey(key)) {
                cache[key] = Entry(value)
                return@withLock null
            }
            
            // If at capacity, evict LFU entry
            var evictedValue: V? = null
            if (cache.size >= capacity) {
                val lfuKey = cache.entries
                    .minByOrNull { it.value.frequency }
                    ?.key
                if (lfuKey != null) {
                    evictedValue = cache.remove(lfuKey)?.value
                    evictedValue?.let { onEvict?.invoke(lfuKey, it) }
                }
            }
            
            cache[key] = Entry(value)
            evictedValue
        }
    }
    
    /**
     * Remove a key from the cache
     */
    fun remove(key: K): V? {
        return lock.withLock {
            cache.remove(key)?.value
        }
    }
    
    /**
     * Check if a key exists in the cache
     */
    fun contains(key: K): Boolean = cache.containsKey(key)
    
    /**
     * Clear all entries
     */
    fun clear() {
        lock.withLock {
            cache.clear()
        }
    }
    
    /**
     * Get current size
     */
    val size: Int get() = cache.size
    
    /**
     * Get cache statistics
     */
    fun getStats(): CacheStats {
        val hits = hitCount.get()
        val misses = missCount.get()
        return CacheStats(
            size = size,
            capacity = capacity,
            hits = hits,
            misses = misses,
            hitRate = if (hits + misses > 0) hits.toDouble() / (hits + misses) else 0.0
        )
    }
}

/**
 * Two-Level Cache combining LRU and TTL
 * L1: Fast access, limited size (LRU)
 * L2: Longer storage with TTL
 */
class TwoLevelCache<K, V>(
    lruCapacity: Int = 100,
    ttlCapacity: Int = 1000,
    defaultTtlMs: Long = 300_000L // 5 minutes
) {
    private val l1Cache = LRUCache<K, V>(lruCapacity)
    private val l2Cache = TTLCache<K, V>(defaultTtlMs, ttlCapacity)
    
    /**
     * Get from L1 first, then L2
     */
    fun get(key: K): V? {
        // Try L1 first
        l1Cache.get(key)?.let { return it }
        
        // Try L2
        return l2Cache.get(key)?.also { 
            // Promote to L1 on hit
            l1Cache.put(key, it)
        }
    }
    
    /**
     * Put in both levels
     */
    fun put(key: K, value: V) {
        l1Cache.put(key, value)
        l2Cache.put(key, value)
    }
    
    /**
     * Put with custom TTL
     */
    fun put(key: K, value: V, ttlMs: Long) {
        l1Cache.put(key, value)
        l2Cache.put(key, value, ttlMs)
    }
    
    /**
     * Remove from both levels
     */
    fun remove(key: K): Boolean {
        val fromL1 = l1Cache.remove(key) != null
        val fromL2 = l2Cache.remove(key) != null
        return fromL1 || fromL2
    }
    
    /**
     * Clear both levels
     */
    fun clear() {
        l1Cache.clear()
        l2Cache.clear()
    }
    
    /**
     * Get combined stats
     */
    fun getStats(): TwoLevelCacheStats {
        return TwoLevelCacheStats(
            l1Stats = l1Cache.getStats(),
            l2Stats = l2Cache.getStats()
        )
    }
}

/**
 * Cache statistics data class
 */
data class CacheStats(
    val size: Int,
    val capacity: Int,
    val hits: Long,
    val misses: Long,
    val hitRate: Double
)

/**
 * TTL Cache statistics data class
 */
data class TTLCacheStats(
    val size: Int,
    val maxSize: Int,
    val hits: Long,
    val misses: Long,
    val hitRate: Double,
    val evictions: Long,
    val defaultTtlMs: Long
)

/**
 * Two-level cache statistics
 */
data class TwoLevelCacheStats(
    val l1Stats: CacheStats,
    val l2Stats: TTLCacheStats
)

/**
 * Simple memoization cache for function results
 */
class MemoCache<K, V>(
    private val compute: (K) -> V
) {
    private val cache = ConcurrentHashMap<K, V>()
    
    fun get(key: K): V = cache.computeIfAbsent(key, compute)
    
    fun clear() = cache.clear()
    
    val size: Int get() = cache.size
    
    fun contains(key: K): Boolean = cache.containsKey(key)
}